cEP-0028.md

```diff
@@ -0,0 +1,479 @@
+| Type     | Process                               |
+
+## Abstract
+
+This cEP describes the optimizations that need to be carried out
+to improve the performance of the coala based language server.
+
+## Introduction
+
+Language servers are the tools used to serve code editors with
+functionalities such as linting, code fixing, analysis etc in a
+standalone and independent manner. coala as a code style checker,
+linter and code fixer could be a powerful tool to be integrated
+into the code editor.
+
+Code Editing is a continuous process and the editors are expected
+to provide feedback on the fly, coala on the other hand was primarily
+designed as an analysis tool to be used after an incremental update to
+the codebase. Thus, the current implementation of a coala language
+server is severely limited by coala's performance.
+
+The performance of the server can be improved by primarily optimizing
+the coala and language server interface and improving the bear
+configuration options to avoid long running bears.
+
+## Proposal
+
+The optimization process focuses to maintain minimal changes to coala
+core and rather optimize the coala and language server interface. The
+following are the proposed ideas.
+
+- Section Tagging
+
+  Some sections are not often required to be executed on every cycle of
+  coala analysis especially during use cases such as this. The parameter
+  set can thus be extended to support a parameter named `tags`. Since
+  these parameters are not associated with any particular bear but rather
+  control the effect of various sections during executing coala, no support
+  extension for bears would be necessary. These parameters enable support
+  for hooking a section for a particular tag type. For example:
+
+  ```
+  [longrunning]
+  tags = save
+  enabled = True
+  bears = SomeLongRunningBear
+  ```
+
+  This kind of a configuration will enable this section during analysis
+  cycles of coala via a language server when a save request is raised
+  while not impacting any untagged regular invocations.
+
+  ```
+  [longrunning]
+  tags = save
+  files = *.py
+  bears = SomeLongRunningBear
+
+  [longrunning.some]
+  tags += change
+  bears = SomeLongRunningBear
+  ```
+
+  This example shows a case where the sections inherit the tags and
+  thereby `longrunning.some` runs on both `save` and `change` tagged
+  execution cycles.
+
+  To enable tagged executional support, coala-core needs to be extended
+  and at least support a new optional INPUT cli argument taking list of
+  tags to be executed. The execution of the sections will be based upon
+  the sections being tagged as at least one of the tags being invoked.
+  This extension of support for `tags` would require at least the
+  following changes:
+
+  - `coalib.parsing.filters.TagSectionFilter`
+
+    Adds a new section filter to the to check if the
+    section should be enabled for a particular tag
+    based invocation.
+
+    ```python
+    def tag_filter(section, args):
```

Comment:
just thinking, does it make sense to make this a method of `Section`?

Reply:
I think that would bloat the `Section` class, instead we are currently using Section Filters, this would enabled us to keep filtering as a general option and allow us to make Section Filters a general concept.

The Section filters framework can later used with targets to support other kinds of filtering.